### ISSUE TITLE
meta:refresh_inventory does not resolve EC2 dynamic inventory properly.

### ISSUE THREAD

**Issue by gitolathe** [ansible/ansible#15115]
##### Issue Type:

- Bug Report
##### Ansible Version:

```
ansible 2.0.1.0
  config file = /home/ola_theander/GIT/ansible_cloud/ansible.cfg
  configured module search path = Default w/o overrides
```
##### Ansible Configuration:

```
[defaults]
host_key_checking = False
# Shorten the contol path to avoid "too long for Unix domain socket" error.
[ssh_connection]
control_path = %(directory)s/%%h-%%p-%%r
```
##### Environment:

N/A
##### Summary:

The included playbook create two instances on AWS EC2 and then wait for them to spin up, performs the shell task `shell: ./ec2.py --refresh-cache` to update the cached dynamic inventory followed by a `meta: refresh_inventory` task to update the in-memory inventory.

The playbook ran as:

```
ansible-playbook -i ec2.py -vvvv main.yml
```

The problem is that apparently the `meta: refresh_inventory` doesn't resolve the host properly. In the subsequent _Install Java on tagged hosts_ step the `hosts:` is properly matched from the inventory but the host DNS name is not resolved. Instead Ansible tries to use the `hostvars` key `ec2_54_74_255_98_eu_west_1_compute_amazonaws_com` for the SSH connection which obviously won't work.

If I break out the install-java step into a separate playbook thus not needing the in-memory refresh done by `meta: refresh_inventory` the Java install step resolves the host just fine. I.e.

```
ansible-playbook -i ec2.py -vvvv main.yml
ansible-playbook -i ec2.py -vvvv install_java.yml
```

The disadvantage with this approach is that it require a manual two-step launch of the two playbooks.
##### Steps To Reproduce:

To reproduce; use a playbook that:
1. create one or more EC2 instances.
2. use a shell command `./ec2.py --refresh-cache` to refresh the cache.
3. refresh the in-memory inventory using `meta: refresh_inventory`.
4. have a subsequent playbook section doing an operation on a group of hosts matching any of the tags generated by the `ec2.py` inventory script.

The details for the `aws` role are omitted since the problem is not really the creation of the EC2 instances but how the inventory is resolved by `meta: refresh_inventory`.

```

---
- name: Create EC2 instances
  hosts: localhost
  connection: local
  gather_facts: false
  # Definition of EC2 instances inspired by https://github.com/chrismeyersfsu/playbook-ec2_properties.
  vars:
    ec2_instances:
      - name: "java1"
        exact_count: 2
        count_tag: "java"
        tags: {Name: java1, java: ''}
      - name: "machine2"
        exact_count: 1
        count_tag: "foo1"
        tags: {Name: demo2, foo1: '', foo3: ''}
  roles:
    - aws
  tasks:
    - name: Wait for SSH
      wait_for:
        host: "{{ item['instances'][0]['public_ip'] }}"
        port: 22
        delay: 10
        timeout: 320
        state: started
      with_items: "{{ ec2.results }}"

    - name: Refresh the ec2.py cache
      shell: ./ec2.py --refresh-cache

    - name: Refresh inventory
      meta: refresh_inventory

# Install Java on all instances tagged with "java".
- name: Install Java on tagged hosts
  hosts: tag_java
  gather_facts: true
  remote_user: ubuntu
  roles:
    - java
```

Output from `ec2.py --list` is

```
{
  "_meta": {
    "hostvars": {
      "ec2_54_74_255_98_eu_west_1_compute_amazonaws_com": {
        "ansible_ssh_host": "ec2-54-74-255-98.eu-west-1.compute.amazonaws.com", 
        "ec2__in_monitoring_element": false, 
        "ec2_ami_launch_index": "0", 
        "ec2_architecture": "x86_64", 
        ...
        ..
  "tag_java": [
    "ec2_54_74_255_98_eu_west_1_compute_amazonaws_com"
  ], 
  ...
  ..
}
```

##### Expected Results:

The expected result is that `meta: refresh_inventory` should properly resolve the `ansible_ssh_host` just as if the inventory was provided on the command-line without an in-memory refresh.
##### Actual Results:

Ansible tries to resolve the host using the  `hostvars` key `ec2_54_74_255_98_eu_west_1_compute_amazonaws_com` which obviously won't work since it's not a valid host. 

```
TASK [setup] *******************************************************************
<ec2_54_74_255_98_eu_west_1_compute_amazonaws_com> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<ec2_54_74_255_98_eu_west_1_compute_amazonaws_com> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Kbd
InteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -
o ConnectTimeout=10 -o ControlPath=/home/ola_theander/.ansible/cp/%h-%p-%r -tt ec2_54_74_255_98_eu_west_1_compute_amazonaws_com '/bin/sh -c '"'"'( umas
k 22 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1458725458.0-274823220393697 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1458725458.0-274
823220393697 `" )'"'"''
fatal: [ec2_54_74_255_98_eu_west_1_compute_amazonaws_com]: UNREACHABLE! => {"changed": false, "msg": "SSH encountered an unknown error. The output was:
\nOpenSSH_6.9p1 Ubuntu-2ubuntu0.1, OpenSSL 1.0.2d 9 Jul 2015\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config l
ine 19: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug1: Control socket \"/home/ola_theander/.ansible/cp/ec2_54_74_255_98_
eu_west_1_compute_amazonaws_com-22-ubuntu\" does not exist\r\nssh: Could not resolve hostname ec2_54_74_255_98_eu_west_1_compute_amazonaws_com: Name or
 service not known\r\n", "unreachable": true}
        to retry, use: --limit @main.retry
```

**Comment by pgporada** [comment]
I get the same sort of error when using the following ec2.ini.

```
[ec2]
regions = us-east-1
regions_exclude = us-gov-west-1,cn-north-1
destination_variable = private_dns_name
vpc_destination_variable = private_ip_address
route53 = True
all_instances = False
all_rds_instances = False
all_elasticache_replication_groups = False
all_elasticache_clusters = False
all_elasticache_nodes = False
cache_path = ~/.ansible/tmp
cache_max_age = 0
nested_groups = False
replace_dash_in_groups = True
expand_csv_tags = False
group_by_instance_id = True
group_by_region = True
group_by_availability_zone = True
group_by_ami_id = True
group_by_instance_type = True
group_by_key_pair = True
group_by_vpc_id = True
group_by_security_group = True
group_by_tag_keys = True
group_by_tag_none = True
group_by_route53_names = True
group_by_rds_engine = True
group_by_rds_parameter_group = True
group_by_elasticache_engine = True
group_by_elasticache_cluster = True
group_by_elasticache_parameter_group = True
group_by_elasticache_replication_group = True
```

Output from /etc/ansible/ec2.py --list looks like

```
{
  "_meta": {
    "hostvars": {
      "10_0_6_7": {
        "ansible_ssh_host": "10.0.6.7", 
        ...
       },
  ...
  ..
}
```

Ansible can't hit 10_0_6_7 because it's not a valid address. I can work around this though.

**Comment by brandond** [comment]
This is fixed by https://github.com/ansible/ansible/pull/15890, but you should also really get the fix from https://github.com/ansible/ansible/issues/14435 or other things will break afterwards as well.

**Comment by ocervell** [comment]
I'm using Ansible 2.5.0 and I'm still running into the same issue.
@brandond #15890 has been merged, but #14435 seems to have been forgotten ?
